# APRORI
***
# <center> Tunggal Manda Ary Triyono
# ---------------------------------------------------------------------------------------------------------------

# Input produk yang sudah dibeli customer

In [1]:
PRODUCT = 'Soap'

# Import Library

In [2]:
import pandas as pd
from apyori import apriori

# Import Dataset

In [3]:
df = pd.read_csv('../data/form2.csv')
df.head(3)

HP  Racket      Watch
0      HP  Camera      Watch
1   Watch  Camera  Music Pad
2  Camera   Watch      Mouse

# Menjadikan dataset menjadi format list

In [4]:
df_list = []
for i in range(0, 23):
    df_list.append([str(df.values[i,j]) for j in range(0,3)])

df_list

[['HP', 'Camera', 'Watch'],
 ['Watch', 'Camera', 'Music Pad'],
 ['Camera', 'Watch', 'Mouse'],
 ['HP', 'Watch', 'Music Pad'],
 ['Watch', 'Racket', 'Camera'],
 ['HP', 'Camera', 'Watch'],
 ['Watch', 'Camera', 'Music Pad'],
 ['Racket', 'Soap', 'Guitar'],
 ['Racket', 'Camera', 'Guitar'],
 ['Camera', 'Bag', 'nan'],
 ['Music Pad', 'Guitar', 'Camera'],
 ['Camera', 'Watch', 'nan'],
 ['Guitar', 'Camera', 'Music Pad'],
 ['Camera', 'Watch', 'Music Pad'],
 ['Camera', 'Racket', 'Guitar'],
 ['Guitar', 'Camera', 'Watch'],
 ['Guitar', 'Watch', 'nan'],
 ['Camera', 'Watch', 'nan'],
 ['Soap', 'Bag', 'Guitar'],
 ['Router', 'Bag', 'nan'],
 ['Watch', 'Racket', 'Soap'],
 ['Music Pad', 'Soap', 'Watch'],
 ['Mouse', 'Camera', 'Soap']]

***
Dataset sebelumnya berformat **dataframe**. untuk digunakan dalam proses apriori menggunakan library apyori, dataset perlu diubah kedalam bentuk **list** agar hasilnya dapat diolah dengan mudah
***

# Menjalankan apriori

In [5]:
frequent_items = apriori(df_list, min_support=0.05, min_confidence=0.05, min_lift=1.5)

***
Menjalankan algoritma apriori dengan nilai 
<br>**Support** : 0.03
<br>**Confidence** : 0.05
<br>**Lift** : 1.5
<br>nilai ini dapat diubah-ubah. **semakin kecil** nilainya, **pasangan item** yang dihasilkan akan **semakin banyak**
***

# Menyimpan hasil apriori dalam list

In [6]:
rules = list(frequent_items)
rules

[RelationRecord(items=frozenset({'nan', 'Bag'}), support=0.08695652173913043, ordered_statistics=[OrderedStatistic(items_base=frozenset({'Bag'}), items_add=frozenset({'nan'}), confidence=0.6666666666666666, lift=3.0666666666666664), OrderedStatistic(items_base=frozenset({'nan'}), items_add=frozenset({'Bag'}), confidence=0.4, lift=3.066666666666667)]),
 RelationRecord(items=frozenset({'Racket', 'Guitar'}), support=0.13043478260869565, ordered_statistics=[OrderedStatistic(items_base=frozenset({'Guitar'}), items_add=frozenset({'Racket'}), confidence=0.375, lift=1.725), OrderedStatistic(items_base=frozenset({'Racket'}), items_add=frozenset({'Guitar'}), confidence=0.6, lift=1.725)]),
 RelationRecord(items=frozenset({'HP', 'Watch'}), support=0.13043478260869565, ordered_statistics=[OrderedStatistic(items_base=frozenset({'HP'}), items_add=frozenset({'Watch'}), confidence=1.0, lift=1.6428571428571428), OrderedStatistic(items_base=frozenset({'Watch'}), items_add=frozenset({'HP'}), confidence=0.

***
Hasilnya terlihat kurang rapi dan sulit dibaca sehingga perlu dirapikan menjadi berbentuk tabel(dataframe)
***

# Total rules yang dihasilkan

In [7]:
len(rules)

6

***
Pasangan Item yang dihasilkan pada apriori ini berjumlah 6
***

# Menampilkan hasil rules menjadi Dataframe

In [8]:
result = []
for item in rules:
    pair = item[0]
    items = [x for x in pair]
    
    value0 = str(items[0])
    value1 = str(items[1])
    #second index of the inner list
    value2 = str(item[1])[:7]
    
    value3 = str(item[2][0][2])[:7]
    value4 = str(item[2][0][3])[:7]
    
    rows = (value0,value1,value2,value3,value4)
    result.append(rows)
col = ['barang1','barang2', 'support', 'confidence', 'lift']
final_result = pd.DataFrame.from_records(result,columns = col)

final_result

barang1 barang2  support confidence     lift
0     nan     Bag  0.08695    0.66666  3.06666
1  Racket  Guitar  0.13043      0.375    1.725
2      HP   Watch  0.13043        1.0  1.64285
3    Soap  Racket  0.08695        0.4     1.84
4  Racket  Camera  0.08695        0.4     1.84
5      HP  Camera  0.08695        1.0  1.64285

***
Data diatas adalah data ke 6 pasangan item yang dibeli pelanggan
<br>
<br>**Support** : seberapa sering itemset muncul di dataset
<br>**confidence** : seberapa sering rules terbukti benar
<br>**lift** : rasio dari confidence untuk persentase jika kejadian mengandung a maka mengandung b
***

# Hasil prediksi barang yang dibeli selanjutnya

In [9]:
PREDICT = final_result.where(final_result['barang1']==PRODUCT)
PREDICT = PREDICT.dropna().sort_values(by='confidence', ascending=False)
a = PREDICT['barang2'].values
print('Barang yang dibeli user setelah membeli', PRODUCT ,'adalah : ',a)

Barang yang dibeli user setelah membeli Soap adalah :  ['Racket']


# KESIMPULAN
Algoritma **Apriori** digunakan untuk menentukan item yang dibeli selanjutnya oleh customer setelah customer membeli barang tertentu
<br>Pada Case ini dimana user telah membeli soap maka prediksi barang yang selanjutnya dibeli oleh user adalah Racket. 
<br>itemset Soap dan Racket **support**nya adalah **0.08695**, artinya persentase kejadian Soap dan Racket dibeli adalah 8,695% 
<br>itemset Soap dan Racket **confidence**nya adalah **0.4**, artinya persentase kejadian user membeli Soap dan membeli Racket adalah 40%
<br>itemset Soap dan Racket **lift**nya adalah **1.84**, artinya rasio kejadian user membeli Soap dan membeli Racket adalah 1.84 lebih banyak dibandingkan user hanya membeli Soap